python/lsst/afw/table/catalogMatches.py

```diff
@@ -0,0 +1,164 @@
+# 
+# LSST Data Management System
+# Copyright 2008-2015 AURA/LSST.
+# 
+# This product includes software developed by the
+# LSST Project (http://www.lsst.org/).
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+# 
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+# 
+# You should have received a copy of the LSST License Statement and 
+# the GNU General Public License along with this program.  If not, 
+# see <https://www.lsstcorp.org/LegalNotices/>.
+#
+from __future__ import absolute_import
+
+import os.path
+
+from .tableLib import BaseCatalog, SimpleCatalog, SourceCatalog, SimpleTable, SourceTable, \
+                      Schema, ReferenceMatch
+from lsst.utils import getPackageDir
+
+__all__ = ["copySchema", "copyCatalog", "matchesToCatalog", "matchesFromCatalog"]
+
+def copySchema(schema, target, targetPrefix=None, sourcePrefix=None):
+    """Return a deep copy the provided schema
```

Comment:
Good points.  I've indicated that the target is modified in place in the docstring and removed the return statement.  On that topic, do we have a preference on how to "return" from functions with no return value?  I've seen no `return` statements, naked `return`, explicit `return None`.  Is there any reason to prefer one over the others?

Also, thanks for pointing out the benefits of using the `SchemaMapper` here (and for writing the code to do it!  This will serve as a nice guideline to clean up the mappings I'm working on to map between HSC vs. LSST schemas for run comparisons).  The only real difference I see between your versions and the old ones is the handling of "existing" keys in the target schema.  In the latter, if a key was already present in the target schema it was silently skipped over.  Now, I believe a name collision will cause the SchemaMapper to throw (since we are using the default `doReplace=false`).  I actually think this is preferred as the user should probably not be trying to copy between two schemas that have some names in common (without having accommodated that with the optional prefix parameters).